torch/csrc/distributed/rpc/rref.cpp

```diff
@@ -54,55 +56,60 @@ const RRefId& RRef::id() const {
   return rrefId_;
 }
 
-at::IValue RRef::fork() const {
+RRefForkData RRef::fork() const {
   return RRefForkData(
       ownerId_, rrefId_, RRefContext::getInstance()->genRRefId()
-  ).toIValue();
+  );
   // NB: does not support sharing RRefs between users
   // TODO: notify the owner
 }
 
 //////////////////////////  UserRRef  /////////////////////////////////////
 
-
-UserRRef::UserRRef(
+template <typename T>
+UserRRef<T>::UserRRef(
     worker_id_t ownerId, const RRefId& rrefId, const ForkId& forkId)
     : RRef(ownerId, rrefId), forkId_(forkId) {
   AT_ASSERT(!(forkId_ == rrefId_),
       "User RRef's fork ID should not be the same as its rref Id");
-  if (RRefContext::getInstance()->getWorkerId() == rrefId_.createdOn_) {
-    // creator user, notify owner.
-    auto& agent = RRefContext::getInstance()->agent();
-    agent->send(
-        agent->getWorkerId(ownerId_),
-        ScriptRRefCreate(
-            RRefForkData(ownerId_, rrefId_, forkId_).toIValue()
-        ).toMessage());
-  } else {
-    AT_ERROR("Does not support sharing RRefs between users yet");
-  }
+  // Do nothing,
+  // (1) If this UserRRef is shared from another UserRRef x, x should
+  // notified the owner on my behalf.
+  // (2) If this UserRRef is shared from the OwnerRRef, the OwnerRRef already
+  // knows this UserRRef.
+  // (3) If this the creator UserRRef, ScriptRemoteCall will properly notify
+  // the owner.
 }
 
-UserRRef::~UserRRef() {
+template <typename T>
+UserRRef<T>::~UserRRef() {
   auto& ctx = RRefContext::getInstance();
   if (ctx->getWorkerId() != ownerId_) {
     ctx->agent()->send(
```

Comment:
need to handle rpc error here. (e.g. print specific error messager containing the RREfId etc, and terminate).